apollo-router/src/axum_http_server_factory.rs

```diff
@@ -0,0 +1,1031 @@
+        );
+    }
+
+    Ok::<_, String>(res)
+}
+
+async fn redirect_or_run_graphql_operation(
+    host: Host,
+    RawQuery(query): RawQuery,
+    headers: HeaderMap,
+    OriginalUri(uri): OriginalUri,
+    Extension(service): Extension<BufferedService>,
+    display_landing_page: bool,
+) -> impl IntoResponse {
+    if headers.get("accept").map(prefers_html).unwrap_or_default() && display_landing_page {
+        return display_home_page().into_response();
+    }
+
+    if query.is_some() {
+        if let Ok(request) =
+            graphql::Request::from_urlencoded_query(query.expect("checked before;qed"))
+        {
+            return run_graphql_request(service, Method::GET, request, headers, uri, host)
+                .await
+                .into_response();
+        }
+    }
+
+    (StatusCode::BAD_REQUEST, "Invalid Graphql request").into_response()
+}
+
+async fn run_graphql_operation(
+    host: Host,
+    OriginalUri(uri): OriginalUri,
+    Json(request): Json<graphql::Request>,
+    Extension(service): Extension<BufferedService>,
+    header_map: HeaderMap,
+) -> impl IntoResponse {
+    run_graphql_request(service, Method::POST, request, header_map, uri, host)
+        .await
+        .into_response()
+}
+
+fn display_home_page() -> Html<Bytes> {
+    let html = Bytes::from_static(include_bytes!("../resources/index.html"));
+    Html(html)
+}
+
+async fn health_check() -> impl IntoResponse {
+    Json(json!({ "status": "pass" }))
+}
+
+// graphql_request is traced at the info level so that it can be processed normally in apollo telemetry.
+#[tracing::instrument(skip_all,
+    level = "info"
+    name = "graphql_request",
+    fields(
+        query = %request.query.clone().unwrap_or_default(),
+        operation_name = %request.operation_name.clone().unwrap_or_else(|| "".to_string()),
+        client_name,
+        client_version
+    )
+)]
+async fn run_graphql_request(
+    service: Buffer<
+        BoxService<
+            http_compat::Request<graphql::Request>,
+            http_compat::Response<ResponseBody>,
+            BoxError,
+        >,
+        http_compat::Request<graphql::Request>,
+    >,
+    method: Method,
+    request: graphql::Request,
+    header_map: HeaderMap,
+    uri: Uri,
+    Host(host): Host,
+) -> impl IntoResponse {
+    if let Some(client_name) = header_map.get("apollographql-client-name") {
+        // Record the client name as part of the current span
+        Span::current().record("client_name", &client_name.to_str().unwrap_or_default());
+    }
+    if let Some(client_version) = header_map.get("apollographql-client-version") {
+        // Record the client version as part of the current span
+        Span::current().record(
+            "client_version",
+            &client_version.to_str().unwrap_or_default(),
+        );
+    }
+
+    match service.ready_oneshot().await {
+        Ok(mut service) => {
+            let uri = format!("http://{}{}", host, uri);
```

Comment:
it would be good to look at the port and the forwarded headers to set the scheme, but let's leave that for a follow up PR

Reply:
Created https://github.com/apollographql/router/issues/835